### PR TITLE
Added Avatars ( #510 )

### PR DESCRIPTION
**Pull request by Doom032**
This pull request adds ( issue #510 )
 - Ability to view avatars of your friends
 - Ability to set your own avatar (left mouse button on avatar)
 - Ability to remove your avatar (right mouse button on avatar and click remove)
 - Avatars being saved locally after receiving them
 - Avatar hashes being calculated and saved and used for comparing if an avatar has changed

Along the way I had to add a lot of code.

Notes:
 - I can't test this well on Windows, because I can't compile it there and mingw's gdb doesn't recognize the utox.exe I cross-compiled from linux for some reason. Anyway, I tested by restarting it 100 times and using all the features and everything seems to work. But if anyone could test it on windows on their own machine that'd be nice.
 - Currently the avatar you choose has to be a PNG under 16KiB. I'd rather have you be able to choose any file size and have it automatically resize but that is something that can be added later(a resize and an image_to_png function will probably be needed)
 - Avatar quality is a bit lower than in qTox, because we don't have nice filtering. This could be added later too.
 - Avatar background is black, while in qTox it seems to automatically makes the background contrast with the image, this means black sillhouettes that are visible in qTox will be pure black images in uTox, automatic contrasting background colors would of course be better
 - This implementation is generally compatible with qTox's, and saves avatars in the same folder and displays the same area of the avatars.

screenshot:
screenshot

### CODE REVIEW

**Comment by irungentoo** [comment]
```
Program received signal SIGSEGV, Segmentation fault.
0x000000000375f2b1 in g_free ()
(gdb) where
#0  0x000000000375f2b1 in g_free ()
#1  0x0000608000009ca0 in ?? ()
#2  0x0000003d8188e5cb in g_private_replace () from /usr/lib64/libglib-2.0.so.0
#3  0x0000003d8b801796 in g_module_symbol () from /usr/lib64/libgmodule-2.0.so.0
#4  0x000000394813a0e0 in ?? () from /usr/lib64/libgtk-x11-2.0.so.0
#5  0x000000394813a11d in ?? () from /usr/lib64/libgtk-x11-2.0.so.0
#6  0x0000003d818578e7 in g_option_context_parse () from /usr/lib64/libglib-2.0.so.0
#7  0x000000394813a3ce in gtk_parse_args () from /usr/lib64/libgtk-x11-2.0.so.0
#8  0x000000394813a429 in gtk_init_check () from /usr/lib64/libgtk-x11-2.0.so.0
#9  0x000000394813a459 in gtk_init () from /usr/lib64/libgtk-x11-2.0.so.0
#10 0x00000000004d4d6e in gtk_load () at ./xlib/gtk.c:224
#11 0x00000000005114eb in main (argc=2, argv=0x7fffffffd858) at ./xlib/main.c:877
(gdb) q
```
Renaming: https://github.com/notsecure/uTox/pull/728/files#diff-592b79e972b9eee0f503ecf3d5b023ceR26 to something else fixes it.

File: main.h [hunk 1]
```diff
@@ -293,6 +308,9 @@ void paste(void);
 void address_to_clipboard(void);
 void openurl(char_t *str);
 void openfilesend(void);
+
+/* use the file chooser to pick an avatar and set it at the user's */
```
Comment (GrayHatter) [inline main.h]: s/at/as

**Comment by GrayHatter** [comment]
@Doom032 consider a version tick in main.h

**Comment by towlie** [comment]
Works well on windows and is compatible with qtox.

**Comment by Doom032** [comment]
@GrayHatter Should I set it to version 0.1.9?

@irungentoo Huh weird, I don't get that error. Maybe that explains why notsecure's code had the g_free's commented out. So I should rename the function pointer g_free to something else? What's a good name? g_free_utox? g_free_it?

@towlie Thanks for testing and I'm glad to hear it works well. And yeah I'm also annoyed by the resizing, but I think it's gonna be a bit more work to get that to be done automatically, so I think it's better to do that separately.

Also, I'm not gonna be as active starting monday because the holiday's over, so if anyone else wants to work on automatic resizing of images and other stuff that'd be possible.

**Comment by towlie** [comment]
@Doom032 It doesn't work that well It seems I restarted utox and it looked like this: 
unbenannt
This happens sometimes with the client avatar too.

**Comment by Doom032** [comment]
@towlie As I said in IRC, I know of that problem but I thought I fixed it. Seems to have to do with a race condition while initializing on Windows. If png_to_image gets called too early on the tox thread it can give the wrong results(usually makes them very white, but can also say the avatar is corrupt while it isn't). The mystery is what in png_to_image has to be initialized before it can safely be called. First I thought it was hdcMem, but in all my tests it got initialized well before the avatars were loaded even if they were loaded the wrong way.

So I thought I practically fixed it by initializing avatars as late as possible in the tox thread initialization, but it seems that only fixed it for me. Now I need to find out what the actual problem is.

**Comment by towlie** [comment]
Windows issue solved by https://github.com/notsecure/uTox/commit/27846a5df895419582bdf9b4f40f61006244fe51

**Pull request merged**